File: Exercism Codes/Java/BookStore.java
import java.util.*;
import java.util.stream.IntStream;

class BookStore {

    double calculateBasketCost(List<Integer> books) {
        Map<Integer, Integer> cnt = getFrequency(books);
        int freq = cnt.values().stream().max(Comparator.comparingInt(x -> x)).orElse(0);
        List<List<Integer>> arr = new ArrayList<>();
        arr.add(new ArrayList<>(Collections.nCopies(freq, 0)));
        List<List<Integer>> combinations = findCombinations(cnt, freq, cnt.keySet().stream().toList(), 0, arr);
        Map<Integer, Double> discounts = Map
                .of(1, 0.0, 2, 0.4, 3, 0.8, 4, 1.6, 5, 2.0);
        return combinations.stream()
                .map(lst -> lst.stream().mapToDouble(b -> b * (8.0 - discounts.get(b)))
                        .sum()).min(Comparator.comparingDouble(a -> a)).orElse(0.0);
    }

    Map<Integer, Integer> getFrequency(List<Integer> books) {
        Map<Integer, Integer> frequency = new HashMap<>();

        for (int x : books) {
            if (frequency.containsKey(x)) {
                frequency.put(x, frequency.get(x) + 1);
            }
            else {
                frequency.put(x, 1);
            }
        }

        return frequency;
    }

    List<List<Integer>> findCombinations(Map<Integer, Integer> cnt, int freq, List<Integer> nums, int num, List<List<Integer>> arr) {
        if (num == nums.size()) {
            return arr;
        }
        else {
            List<List<Integer>> arr2 = new ArrayList<>(arr);
            arr = new ArrayList<>();

            for (List<Integer> x : arr2) {
                for (List<Integer> positions : getCombinations(IntStream.range(0, freq).boxed().toList(), cnt.get(nums.get(num)))) {
                    List<Integer> newArr = new ArrayList<>(x);
                    for (int pos : positions) {
                        newArr.set(pos, newArr.get(pos) + 1);
                    }
                    arr.add(newArr);
                }
            }

            return findCombinations(cnt, freq, nums, num + 1, arr);
        }
    }

    List<List<Integer>> getCombinations(List<Integer> arr, int r) {
        List<List<Integer>> result = new ArrayList<>();
        return backtrack(0, new ArrayList<>(), result, arr, r);
    }

    List<List<Integer>> backtrack(int start, List<Integer> current, List<List<Integer>> result, List<Integer> arr, int r) {
        if (current.size() == r) {
            result.add(new ArrayList<>(current));
            return result;
        }
        for (int i = start; i < arr.size(); i++) {
            current.add(arr.get(i));
            backtrack(i + 1, current, result, arr, r);
            current.removeLast();
        }

        return result;
    }
}

File: Exercism Codes/Python/book_store.py
from collections import Counter
from itertools import combinations


def total(basket: list):
    cnt = Counter(basket)
    freq = max(cnt.values(), default=0)
    combination = find_combinations(dict(cnt), freq, list(cnt.keys()), 0, [[0] * freq])
    discounts = {1: 0.0, 2: 0.4, 3: 0.8, 4: 1.6, 5: 2.0}
    return min(map_combinations(lst, discounts) for lst in combination)

def map_combinations(lst, discounts):
    return sum(100 * b * (8.0 - discounts[b]) for b in lst)


def find_combinations(cnt, freq, nums, num, arr):
    if num == len(nums):
        return arr
    else:
        arr_2 = arr.copy()
        arr = []
        for x in arr_2:
            for positions in combinations(range(freq), cnt[nums[num]]):
                new_arr = x.copy()
                for j in positions:
                    new_arr[j] += 1
                arr.append(new_arr)
        return find_combinations(cnt, freq, nums, num + 1, arr)